packages/flame/lib/src/components/component.dart

```diff
@@ -3,88 +3,284 @@ import 'dart:ui';
+  /// If the component has a parent it will be set here.
   Component? _parent;
 
+  /// Get the current parent of the component, if there is one, otherwise null.
   Component? get parent => _parent;
 
-  /// Render priority of this component. This allows you to control the order in which your components are rendered.
+  /// If the component should be added to another parent once it has been
+  /// removed from its current parent.
+  Component? nextParent;
+
+  late final ComponentSet children = createComponentSet();
+
+  /// Render priority of this component. This allows you to control the order in
+  /// which your components are rendered.
   ///
-  /// Components are always updated and rendered in the order defined by what this number is when the component is added to the game.
-  /// The smaller the priority, the sooner your component will be updated/rendered.
+  /// Components are always updated and rendered in the order defined by what
+  /// this number is when the component is added to the game.
+  /// The smaller the priority, the sooner your component will be
+  /// updated/rendered.
   /// It can be any integer (negative, zero, or positive).
-  /// If two components share the same priority, they will probably be drawn in the order they were added.
+  /// If two components share the same priority, they will probably be drawn in
+  /// the order they were added.
   int get priority => _priority;
   int _priority;
 
   /// Whether this component should be removed or not.
   ///
-  /// It will be checked once per component per tick, and if it is true, BaseGame will remove it.
+  /// It will be checked once per component per tick, and if it is true,
+  /// BaseGame will remove it.
   bool shouldRemove = false;
 
+  /// Returns whether this [Component] is in debug mode or not.
+  /// When a child is added to the [Component] it gets the same [debugMode] as
+  /// its parent has when it is prepared.
+  ///
+  /// Returns `false` by default. Override it, or set it to true, to use debug
+  /// mode.
+  /// You can use this value to enable debug behaviors for your game and many
+  /// components will
+  /// show extra information on the screen when debug mode is activated
+  bool debugMode = false;
+
+  /// How many decimal digits to print when displaying coordinates in the
+  /// debug mode. Setting this to null will suppress all coordinates from
+  /// the output.
+  int? get debugCoordinatesPrecision => 0;
+
+  /// The color that the debug output should be rendered with.
+  Color debugColor = const Color(0xFFFF00FF);
+
+  /// The [debugColor] represented as a [Paint] object.
+  Paint get debugPaint => Paint()
+    ..color = debugColor
+    ..strokeWidth = 1
+    ..style = PaintingStyle.stroke;
+
+  /// Returns a [TextPaint] object with the [debugColor] set as color for the
+  /// text.
+  TextPaint get debugTextPaint => TextPaint(
+        config: TextPaintConfig(
+          color: debugColor,
+          fontSize: 12,
+        ),
+      );
+
```

Comment:
Both these should be memoized to avoid creating instances on every access

Reply:
Good idea, fixed with `ValueCache`